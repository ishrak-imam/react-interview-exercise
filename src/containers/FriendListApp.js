import React, { Component } from 'react'
import styles from './FriendListApp.css'
import { connect } from 'react-redux'

import { addFriend, goToPage } from '../actions/FriendsActions'
import { FriendList, AddFriendInput, Pagination } from '../components'

export class FriendListApp extends Component {
  constructor (props) {
    super(props)

    this.goToPage = this.goToPage.bind(this)
  }

  goToPage (page) {
    this.props.goToPage(page)
  }

  render () {
    const {
      friendlist: {
        ids,
        pagination: { currentPage, pages }
      },
      addFriend
    } = this.props

    const page = pages[currentPage]
    const pageNumbers = Object.keys(pages)

    let friendIds = ids
    if (page) friendIds = page.ids

    return (
      <div className={styles.friendListApp}>
        <h1>The FriendList</h1>
        <AddFriendInput addFriend={addFriend} />
        <FriendList friendIds={friendIds} />
        {
          page && ids.length > 2 &&
          <Pagination currentPage={currentPage} pages={pageNumbers} goToPage={this.goToPage} />
        }
      </div>
    )
  }
}

function mapStateToProps (state) {
  return state
}

export default connect(mapStateToProps, {
  addFriend,
  goToPage
})(FriendListApp)
