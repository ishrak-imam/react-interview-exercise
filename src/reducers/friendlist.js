import * as types from '../constants/ActionTypes'

const ITEMS_PER_PAGE = 3

const data = [
  {
    name: 'Theodore Roosevelt',
    starred: true,
    id: 0,
    gender: 'M'
  },
  {
    name: 'Abraham Lincoln',
    starred: false,
    id: 1,
    gender: 'M'
  },
  {
    name: 'George Washington',
    starred: false,
    id: 2,
    gender: 'M'
  }
]

const normalizeList = list => {
  return list.reduce((obj, item, index) => {
    obj.map[item.id] = item
    obj.ids.push(index)
    return obj
  }, { map: {}, ids: [] })
}

const chunkList = (list, size) => {
  let chunkedList = []
  const noOfChunks = Math.ceil(list.length / size)
  for (let i = 0; i < noOfChunks; i++) {
    chunkedList.push(list.slice(i * size, (i + 1) * size))
  }
  return chunkedList
}

const getPages = ids => {
  const chunkedIds = chunkList(ids, ITEMS_PER_PAGE)
  return chunkedIds.reduce((pages, chunk, index) => {
    pages[index + 1] = { ids: chunk }
    return pages
  }, {})
}

const { map, ids } = normalizeList(data)

const initialState = {
  friends: map,
  ids,
  pagination: {
    itemsPerPage: ITEMS_PER_PAGE,
    currentPage: '1',
    pages: getPages(ids)
  }
}

export default function friends (state = initialState, action) {
  switch (action.type) {
    case types.ADD_FRIEND:
      let ids = state.ids
      let friends = { ...state.friends, [ids.length]: { ...action.friend, id: ids.length } }
      ids.push(ids.length)

      return {
        ...state,
        friends,
        ids,
        pagination: {
          ...state.pagination,
          pages: getPages(ids)
        }
      }

    case types.DELETE_FRIEND:
      ids = state.ids.filter(id => id !== action.id)
      friends = Object.assign({}, state.friends)
      delete friends[action.id]

      // let { [action.id]: deleted, ...afterDelete } = friends

      return {
        ...state,
        friends,
        ids,
        pagination: {
          ...state.pagination,
          pages: getPages(ids)
        }
      }

    case types.STAR_FRIEND:
      friends = state.friends
      let friend = friends[action.id]
      return {
        ...state,
        friends: {
          ...friends,
          [action.id]: {
            ...friend,
            starred: true
          }
        }
      }

    case types.GO_TO_PAGE:
      return {
        ...state,
        pagination: {
          ...state.pagination,
          currentPage: action.page
        }
      }

    default:
      return state
  }
}
