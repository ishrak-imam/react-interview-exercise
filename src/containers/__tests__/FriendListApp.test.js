import React from 'react'
import ShallowRenderer from 'react-test-renderer/shallow'
import { FriendListApp } from '../FriendListApp'

import { initialState } from '../../reducers/friendlist'

test('FriendListApp component', () => {
  const renderer = new ShallowRenderer()
  const component = renderer.render(
    <FriendListApp
      addFriend={() => {}}
      goToPage={() => {}}
      friendlist={initialState}
    />
  )
  expect(component).toMatchSnapshot()
})
