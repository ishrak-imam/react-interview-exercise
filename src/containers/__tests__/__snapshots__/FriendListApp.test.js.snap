exports[`test FriendListApp component 1`] = `
<div
  className={undefined}>
  <h1>
    The FriendList
  </h1>
  <AddFriendInput
    addFriend={[Function]} />
  <FriendList
    friendIds={
      Array [
        "0",
        "1",
        "2",
      ]
    } />
  <Pagination
    currentPage="1"
    goToPage={[Function]}
    pages={
      Array [
        "1",
      ]
    } />
</div>
`;
